src/commands/force/source/beta/pull.ts

```diff
@@ -0,0 +1,185 @@
+    });
+
+    this.ux.startSpinner('Loading source tracking information');
+    this.tracking = await SourceTracking.create({
+      org: this.org,
+      project: this.project,
+    });
+
+    await this.tracking.ensureRemoteTracking(true);
+    this.ux.setSpinnerStatus('Checking for conflicts');
+
+    if (!this.flags.forceoverwrite) {
+      processConflicts(await this.tracking.getConflicts(), this.ux, messages.getMessage('sourceConflictDetected'));
+    }
+  }
+
+  protected async doDeletes(): Promise<void> {
+    this.ux.setSpinnerStatus('Checking for deletes from the org and updating source tracking files');
+    const changesToDelete = await this.tracking.getChanges<SourceComponent>({
+      origin: 'remote',
+      state: 'delete',
+      format: 'SourceComponent',
+    });
+    this.deleteFileResponses = await this.tracking.deleteFilesAndUpdateTracking(changesToDelete);
+  }
+
+  protected async updateTrackingFilesWithRetrieve(): Promise<void> {
+    this.ux.setSpinnerStatus('Updating source tracking files');
+
+    const successes = this.retrieveResult
+      .getFileResponses()
+      .filter((fileResponse) => fileResponse.state !== ComponentStatus.Failed);
+
+    await Promise.all([
+      // commit the local file successes that the retrieve modified
+      this.tracking.updateLocalTracking({
+        files: successes.map((fileResponse) => fileResponse.filePath).filter(Boolean),
+      }),
+      this.tracking.updateRemoteTracking(
+        successes.map(({ state, fullName, type, filePath }) => ({ state, fullName, type, filePath })),
+        true // skip polling because it's a pull
+      ),
+    ]);
+  }
+
+  protected async retrieve(): Promise<void> {
+    const componentSet = new ComponentSet();
+    (
+      await this.tracking.getChanges<ChangeResult>({
+        origin: 'remote',
+        state: 'nondelete',
+        format: 'ChangeResult',
+      })
+    ).map((component) => {
+      if (component.type && component.name) {
+        componentSet.add({
+          type: component.type,
+          fullName: component.name,
+        });
+      }
+    });
+
+    if (componentSet.size === 0) {
+      return;
+    }
+    componentSet.sourceApiVersion = await this.getSourceApiVersion();
+    if (this.getFlag<string>('apiversion')) {
+      componentSet.apiVersion = this.getFlag<string>('apiversion');
+    }
+
+    const mdapiRetrieve = await componentSet.retrieve({
+      usernameOrConnection: this.org.getUsername(),
+      merge: true,
+      output: this.project.getDefaultPackage().path,
+    });
+
+    this.ux.setSpinnerStatus('Retrieving metadata from the org');
+
+    // assume: remote deletes that get deleted locally don't fire hooks?
+    await this.lifecycle.emit('preretrieve', componentSet.toArray());
+    this.retrieveResult = await mdapiRetrieve.pollStatus(1000, this.getFlag<Duration>('wait').seconds);
+
+    // assume: remote deletes that get deleted locally don't fire hooks?
```

Comment:
> assume: remote deletes that get deleted locally don't fire hooks?
Good question.  Not sure what currently happens in source:pull but match that.  If it's not possible in the legacy version then we can do whatever makes sense.

Reply:
translation: [I was too lazy to try to dig through toolbelt to find out and was hoping someone knew]

Reply:
My reading is that it doesn't emit hooks...the postRetrieveHookInfo is only using the fileProperties from the mdapi response.  I think.